Use loglevel module for JS logging

Mr0grog opened the pull request:
This addresses #201 and allows us to turn IE 10 back on for testing.

There are other ways to handle this and modules we could use besides [loglevel](https://www.npmjs.com/package/loglevel), but I think it’s a good, simple one (and I have experience with it). Happy to close this is others prefer something else.

oderby commented:
How does this interact with logLevel setting in karma.conf? Does karma do it's own override of console? I'm just thinking if there's anything we should do to prevent dumb test failures on CI if someone leaves a `console.debug` call in a test file...

Mr0grog commented:
Oh, I did not think about how this would interact with Karma’s settings. I *believe* it would just be an additional restriction on top of what Karma does (so whichever is stricter gets applied) because I think it’s just restricting whether it repeats what it gets from the browser, but I’ll have to test and see.

This also made me think we should explicitly put the app into debug mode when running tests (as far as I can tell from perusing the codebase, you have to do this yourself by manually setting localStorage right now).

> I'm just thinking if there's anything we should do to prevent dumb test failures on CI if someone leaves a console.debug call in a test file

Hmm, I think we *want* that to fail, though, right? We should just make sure it fails locally in Chrome, too. I’m not sure if there’s a good way to modify the `console` object to throw errors without stepping on loglevel’s toes, but it does have a plugin system we could potentially use to do it.

That said, the simplest thing might be to add ESLint as part of testing and make it throw errors on `console.*` calls.

oderby commented:
> We should just make sure it fails locally in Chrome, too.

Yeah, that's probably the only viable approach.

> That said, the simplest thing might be to add ESLint as part of testing and make it throw errors on console.* calls.

To clarify, you're proposing we make `console.*` calls into errors, because people should use `logLevel` instead, right? (When I first read it, I thought you were proposing we make it an error to do any logging from tests) Assuming so, I like the idea. Seems like it's a [built-in rule already](http://eslint.org/docs/rules/no-console) in ESLint. We should also turn it on for app code, too, right? Finally, it'd be awesome if we could customize the error message/give tips to offenders, like "don't use console, use logLevel instead.".

Mr0grog commented:
> To clarify, you're proposing we make `console.*` calls into errors, because people should use logLevel instead, right? …Seems like it's a built-in rule already in ESLint. We should also turn it on for app code, too, right?

Yep, exactly what I was thinking. And better to just lean on ESLint for this than to somehow make `console.*` calls throw errors.

> Finally, it'd be awesome if we could customize the error message/give tips to offenders, like "don't use console, use logLevel instead."

Good idea. I’m not sure it’s very easy to customize ESLint’s output this way, but I can think of some methods to hack it in.

Mr0grog commented:
Updated with a couple commits to make sure we always show all logged messages (not just `warn` and `error`) in tests. As a side note, I looked into the Karma config file and:

- `logLevel` affects messages coming from Karma itself
- `browserConsoleLogOptions` affects which logs from the browser are passed along by manipulating the `console` object in the browser, but does so in a way that doesn’t conflict with `loglevel`. It only works as an additional level filter on top of what we do with `loglevel`.

I also added in ESLint (run `gulp eslint`), but only using the rule for `console.*` calls (it would be good to add more rules later). The message has been customized to:

```
Please use the `loglevel` module for logging instead of the browser `console` object
```

This makes the implementation look a little complicated, but we still use all the standard options and behave as if you just ran `eslint ./static` from the command line (excepting the customized messages).

The linter also runs as part of the normal `test` task. I think we definitely want that for CI; it seemed like it might be good for local dev, too, but we can split them up again if not.

nickoneill commented:
Who's best to review here? @oderby or @cdoremus?

oderby commented on gulpfile.js:
```diff
@@ -144,7 +144,7 @@ function runKarmaTests ({singleRun, configFile} = {}) {
 }
 
-gulp.task('test', function() {
+gulp.task('test:js-unit', function() {
```
why not just

```javascript
gulp.task('test' ['eslint'], function() {
```

and then no need for line 183?

oderby replied:
Re-reading your comment on the PR thread, I'm guessing you were thinking this way, people who really didn't want to run ESLint for local dev don't have to. But I'd argue that local tests should be subject to same restrictions/requirements as CI, so with the current config (only run no-console rule), it should definitely run for all local tests.

oderby replied:
Also, as the code currently is arranged, if I run `gulp test`, the two tasks (`eslint` and `test:js-unit`) will run in parallel. This means that eslint could fail while js-unit passes, but because js-unit takes longer, my console is filled with the karma reporter output and I don't notice that the linter failed. The process does correctly return a non-zero exit code, but I doubt most would notice that.

Whereas, if you make the js-unit task depend explicitly on the eslint task (like the test:ci task), it sees eslint fails and doesn't even run karma (much better behavior, IMO)

Mr0grog replied:
> I'm guessing you were thinking this way, people who really didn't want to run ESLint for local dev don't have to.

Nah, I was actually thinking more about things like #209 (which I have not had time to look at yet). Mainly this: there will be more “tests” than just the unit tests and the `test` task should just be gathering them all together to run—each type of test should be implemented separately.

> eslint could fail while js-unit passes, but because js-unit takes longer, my console is filled with the karma reporter output and I don't notice that the linter failed.

Sure. I figured it was OK enough that gulp does the job of keeping the output separate, whatever order it was that it came in. We could certainly make them serial and stop as soon as one fails.

Mr0grog replied:
Conveniently solved by https://github.com/5calls/5calls/pull/203#discussion_r103061645 !

oderby commented on gulpfile.js:
```diff
@@ -155,9 +155,32 @@ gulp.task('test:watch', function() {
+    return Promise.reject(
+      new util.PluginError('ESLint', 'Found problems with JS coding style.'));
+  }
+});
```
Shouldn't you return a resolved promise here? (empirically, it seems like gulp works fine without it, but for consistency and correctness, seems wrong to only return a promise on failure)

Mr0grog replied:
Solved this by just throwing the plugin error, which is what I should have originally been doing because it’s a sync task.

oderby commented on static/js/main.js:
```diff
@@ -14,7 +15,7 @@ const appURL = 'https://5calls.org';
 if (!debug) {
-  console.debug = () => {};
+  logger.setDefaultLevel(logger.levels.WARN);
 }
```
It appears (at least for me) the default log level is already WARN. That means that we need to explicitly set the log level to something lower (TRACE) if debug is true. Otherwise, turning on debugging actually has no effect.

Also, did you decide to use `setDefaultLevel` here and `setLevel(level, false)` for the test setup?

Mr0grog replied:
You’re right! It’s been a while since I’ve used `loglevel`—could have sworn it was the other way around (trace by default).

> Also, did you decide to use setDefaultLevel here and setLevel(level, false) for the test setup?

Yes, that’s intentional; the two behave differently (`setDefaultLevel` only sets if no level has been explicitly set before, e.g. on a previous page load, while `setLevel(whatever, false)` *always* sets the level, but does not save it). This ensures that, in tests, we are absolutely always showing all logs, but not in a way that might affect your browser next time you open a page not in a test setting.

Mr0grog replied:
Done.

oderby commented on static/test/setupUnitTests.js:
```diff
@@ -0,0 +1,7 @@
+/**
```
This seems reasonable to me, but @cdoremus may have thoughts?

oderby commented on gulpfile.js:
```diff
@@ -155,9 +155,31 @@ gulp.task('test:watch', function() {
+gulp.task('eslint', function() {
+  const eslint = require('eslint');
+  const linter = new eslint.CLIEngine();
+  const report = linter.executeOnFiles(['./static']);
```
So currently, eslint is run over all files, but only run as part of tests. I believe we should run eslint during normal dev iterations, probably as a dependency of the 'scripts' task? I'm fine filing as an issue and deferring to a separate PR if you'd rather, since the current form at least ensures it will be caught by CI eventually.

Mr0grog replied:
I don’t see linting as something that belongs as a dependency of the `scripts` task. I don’t believe it makes sense to prevent code from compiling and running simply because it’s poor form; I’ll often try an initial pass at something that is ugly and ignores a bunch of good conventions or rules to test whether it works; making it nice happens later (hopefully before committing, but definitely before PR-ing).

I absolutely think not passing a lint test should prevent code from merging, but by the same token, I see it as an acceptance criterion (tests) and not a build criterion.

oderby commented:
That's a good point. I was just thinking of "check that they aren't using
console.*" as what would be checked on build, but forgot that it's part of
a larger (planned to expand) lint process. Given that, I'm fine with this
as is.

On Mon, Feb 27, 2017 at 10:25 AM, Rob Brackett <notifications@github.com>
wrote:

> *@Mr0grog* commented on this pull request.
> ------------------------------
>
> In gulpfile.js
> <https://github.com/5calls/5calls/pull/203#discussion_r103275873>:
>
> >    return runKarmaTests({configFile: 'karma.ci.conf.js'});
>  });
>
> +gulp.task('eslint', function() {
> +  const eslint = require('eslint');
> +  const linter = new eslint.CLIEngine();
> +  const report = linter.executeOnFiles(['./static']);
>
> I don’t see linting as something that belongs as a dependency of the
> scripts task. I don’t believe it makes sense to prevent code from
> compiling and running simply because it’s poor form; I’ll often try an
> initial pass at something that is ugly and ignores a bunch of good
> conventions or rules to test whether it works; making it nice happens later
> (hopefully before committing, but definitely before PR-ing).
>
> I absolutely think not passing a lint test should prevent code from
> merging, but by the same token, I see it as an acceptance criterion (tests)
> and not a build criterion.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/5calls/5calls/pull/203#discussion_r103275873>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABSd9hug7gLNqp7D2tPs3HeKG5hFwewhks5rgxUTgaJpZM4MDUjL>
> .
>
